### PR TITLE
Refactor Mailbox Configuration

### PR DESCRIPTION
**Pull request by nicolas71640**
Change the way of SyncManagers configuration. 
Originally we configure the SyncManagers statically : The AL gives the configuration. This can be avoided as the master should be the one to configure the mailbox sync managers thanks to the eeprom. And the slave only need to check the coherency of this configuration

This PR also export the SII parser to common folder

### CODE REVIEW

**Comment by nicolas71640** [comment]
fix #179 #178

File: lib/slave/include/kickcat/AbstractESC.h [hunk 1]
```diff
@@ -42,7 +47,13 @@ namespace kickcat
 
         void routine();
 
-        void set_mailbox_config(std::vector<SyncManagerConfig> const& mailbox);
+
+        // TODO: Add a comment to say that we use the configs that are passed to this function
```
Comment (nicolas71640) [inline lib/slave/include/kickcat/AbstractESC.h:51]: TODO

File: lib/master/include/kickcat/SIIParser.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        // Categories
+        std::vector<std::string_view> strings;
+        eeprom::GeneralEntry const* general;
+        std::vector<uint8_t> fmmus;
+        std::vector<eeprom::SyncManagerEntry const*> syncManagers;
+        std::vector<eeprom::PDOEntry const*> TxPDO;
+        std::vector<eeprom::PDOEntry const*> RxPDO;
+    };
+
+    SII parseSII(std::vector<uint32_t>& eeprom);
```
Comment (leducp) [inline lib/master/include/kickcat/SIIParser.h:46]: move semantic?

File: lib/src/Mailbox.cc [hunk 3]
```diff
@@ -199,16 +215,71 @@ namespace kickcat::mailbox::response
...
+                return false;
+            }
+
+            mbx_in_  = SYNC_MANAGER_MBX_IN(indexIn, mailboxIn.start_address, mailboxIn.length);
+            mbx_out_ = SYNC_MANAGER_MBX_IN(indexOut, mailboxOut.start_address, mailboxOut.length);
+
+            configured_ = true;
+        }
+        catch (std::exception const& e)
+        {
```
Comment (leducp) [inline lib/src/Mailbox.cc:267]: nothing in the catch?

File: lib/slave/include/kickcat/AbstractESC.h [hunk 4]
```diff
@@ -42,7 +47,13 @@ namespace kickcat
...
+        // when passing from init to preop and that this state change fails if the master ask 
+        // for a mailbox and none has been gave to the abstract esc or inverted.
+    
+        //void set_mailbox_config(Mailbox::response* mailbox, // Max size);
```
Comment (leducp) [inline lib/slave/include/kickcat/AbstractESC.h:55]: commented line

File: lib/src/SIIParser.cc [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        pos += 2;
+
+        for (int32_t i = 0; i < number_of_entries; ++i)
+        {
+            pdo.push_back(reinterpret_cast<eeprom::PDOEntry const*>(pos));
+            pos += sizeof(eeprom::PDOEntry);
+        }
+    }
+
+    SII parseSII(std::vector<uint32_t>& eeprom)
```
Comment (leducp) [inline lib/src/SIIParser.cc:74]: move semantic

**Review by leducp — APPROVED** [review]

File: lib/src/Mailbox.cc [hunk 6]
```diff
@@ -199,16 +215,62 @@ namespace kickcat::mailbox::response
...
+    std::tuple<uint8_t, SyncManager> Mailbox::find_sm(uint16_t controlMode)
+    {
+        for (uint8_t i = 0; i < reg::SM_STATS; i++)
+        {
+            SyncManager sync;
+            esc_->read(reg::SYNC_MANAGER + sizeof(SyncManager) * i, &sync, sizeof(SyncManager));
+            if (sync.control == controlMode)
+            {
+                return std::tuple(i, sync);
+                mbx_in_ = SYNC_MANAGER_MBX_IN(i, sync.start_address, sync.length);
```
Comment (Rdk-T) [inline lib/src/Mailbox.cc:239]: This line will never be called ?

File: lib/src/Mailbox.cc [hunk 7]
```diff
@@ -199,16 +215,62 @@ namespace kickcat::mailbox::response
...
         , mbx_out_{mbx_out}
+        , max_allocated_ram_by_msg_{mbx_in.length}
         , max_msgs_{max_msgs}
     {
+    }
 
+    Mailbox::Mailbox(AbstractESC* esc, uint16_t max_allocated_ram_by_msg, uint16_t max_msgs)
+        : esc_{esc}
+        , max_allocated_ram_by_msg_{max_allocated_ram_by_msg}
+        , max_msgs_{max_msgs}
```
Comment (Rdk-T) [inline lib/src/Mailbox.cc:226]: unused ?

File: unit/slave-t.cc [hunk 8]
```diff
@@ -117,9 +117,9 @@ TEST(Slave, parse_SII)
     };
 
     slave.parseSII();
-    ASSERT_EQ(2, slave.sii.strings.size()); // 2 strings since 0 is unused
-    ASSERT_EQ(4, slave.sii.fmmus.size());
-    ASSERT_EQ(1, slave.sii.syncManagers.size());
+//    ASSERT_EQ(2, slave.sii.strings.size()); // 2 strings since 0 is unused
```
Comment (Rdk-T) [inline unit/slave-t.cc:120]: Commented test ?

**Comment by Rdk-T** [comment]
Before merged the CTT suite shall be run / tested on hardware

File: lib/src/Mailbox.cc (hunk 7 above)
Comment (nicolas71640) [reply lib/src/Mailbox.cc:226]: yes it can be used by caller

File: lib/src/Mailbox.cc (hunk 6 above)
Comment (nicolas71640) [reply lib/src/Mailbox.cc:239]: indeed, useless, removed id

File: unit/slave-t.cc (hunk 8 above)
Comment (nicolas71640) [reply unit/slave-t.cc]: forgotten ! thanks !

File: lib/src/Mailbox.cc [hunk 9]
```diff
@@ -199,16 +215,62 @@ namespace kickcat::mailbox::response
...
+        }
+
+        THROW_ERROR("SyncManager not found");
+    }
+
+    bool Mailbox::configure_sm()
+    {
+        try
+        {
+            auto [indexIn, mailboxIn]   = find_sm(0x22);
```
[the remark this answers is not in the document]
Comment (nicolas71640) [reply lib/src/Mailbox.cc]: Done

File: lib/include/kickcat/Mailbox.h [hunk 10]
```diff
@@ -163,23 +168,29 @@ namespace kickcat::mailbox::response
...
         CoE::Dictionary& getDictionary(){return dictionary_;}
 
+        bool configure_sm(); // Configure sync managers
         void receive(); // Try to receive a message from the ESC
         void process(); // Process a message in the to_process_ queue if any
         void send();    // Send a message in the to_send_ queue if any, keep it in the queue if the ESC is not ready yet
 
     private:
         void replyError(std::vector<uint8_t>&& raw_message, uint16_t code);
+        std::tuple<uint8_t, SyncManager> find_sm(uint16_t controlMode);
```
Comment (trns1997) [inline lib/include/kickcat/Mailbox.h:188]: Same style remark

Comment (nicolas71640) [reply lib/include/kickcat/Mailbox.h]: Moved it to abstract esc, kept snake case as abstractESC function are in snake case... not really consistent but...

File: lib/include/kickcat/Mailbox.h [hunk 11]
```diff
@@ -163,23 +168,29 @@ namespace kickcat::mailbox::response
...
         /// \param max_msgs Max messages allowed simultaneously in the processing queue
         Mailbox(AbstractESC* esc, SyncManagerConfig mbx_in, SyncManagerConfig mbx_out, uint16_t max_msgs = 1);
+
+        Mailbox(AbstractESC* esc, uint16_t max_allocated_ram_by_msg, uint16_t max_msgs = 1);
         ~Mailbox() = default;
 
         void enableCoE(CoE::Dictionary&& dictionary);
         CoE::Dictionary& getDictionary(){return dictionary_;}
 
+        bool configure_sm(); // Configure sync managers
```
[the remark this answers is not in the document]
Comment (nicolas71640) [reply lib/include/kickcat/Mailbox.h]: Changed it to match mailbox consistency

**Review by leducp — APPROVED** [review]

File: lib/master/src/Prints.cc [hunk 12]
```diff
@@ -8,23 +8,23 @@ namespace kickcat
     {
         std::stringstream os;
         os << "\n -*-*-*-*- slave " << std::to_string(slave.address) << " -*-*-*-*-\n";
-        os << "Vendor ID:       " << "0x" << std::setfill('0') << std::setw(8) << std::hex << slave.vendor_id << "\n";
-        os << "Product code:    " << "0x" << std::setfill('0') << std::setw(8) << std::hex << slave.product_code << "\n";
-        os << "Revision number: " << "0x" << std::setfill('0') << std::setw(8) << std::hex << slave.revision_number << "\n";
-        os << "Serial number:   " << "0x" << std::setfill('0') << std::setw(8) << std::hex << slave.serial_number << "\n";
+        os << "Vendor ID:       " << "0x" << std::setfill('0') << std::setw(8) << std::hex << slave.sii.vendor_id << "\n";
```
Comment (trns1997) [inline lib/master/src/Prints.cc:11]: Is it possible to use `std::endl` instead of `\n` for consistency sake?

**Review by trns1997 — APPROVED** [review]

File: lib/master/src/Prints.cc (hunk 12 above)
Comment (nicolas71640) [reply lib/master/src/Prints.cc:11]: This PR has nothing to do with those prints, so I prefer to not add noise on this.

**Review by Rdk-T — APPROVED** [review]

**Pull request merged**